lib/CFRPageActions.jsm

```diff
@@ -0,0 +1,241 @@
+    // CSS transitions from the previous state
+    this.hbox.removeAttribute("state");
+    this.hbox.hidden = false;
+
+    // We can safely get the width of the label even though the recommendation is
+    // collapsed; the label itself remains full width (with its overflow hidden)
+    const [{width}] = this.label.getClientRects();
+    this.hbox.style.setProperty("--cfr-expanded-label-width", `${width}px`);
+
+    if (shouldExpand) {
+      // After one second, expand
+      this.stateTransitionTimeoutIDs.push(this.window.setTimeout(() => {
+        this.hbox.setAttribute("state", "expanded");
+      }, 1000));
+
+      // Five seconds later, collapse again
+      this.stateTransitionTimeoutIDs.push(this.window.setTimeout(() => {
+        this.hbox.setAttribute("state", "collapsed");
+      }, 6000));
+    }
+  }
+
+  hide() {
+    this.hbox.hidden = true;
+    this.hbox.removeAttribute("state");
+  }
+
+  _clearScheduledStateChanges() {
+    while (this.stateTransitionTimeoutIDs.length > 0) {
+      // clearTimeout is safe even with invalid/expired IDs
+      this.window.clearTimeout(this.stateTransitionTimeoutIDs.pop());
+    }
+  }
+
+  // This is called when the popup closes as a result of interaction _outside_
+  // the popup, e.g. by hitting <esc>
+  _popupStateChange(state) {
+    if (["dismissed", "removed"].includes(state)) {
+      this._clearScheduledStateChanges();
+      if (this.hbox.getAttribute("state") === "expanded") {
+        this.hbox.setAttribute("state", "collapsed");
+      }
+    }
+  }
+
+  // This is called when the popup closes as a result of clicking "No thanks"
+  _popupDismissed() {
+    this._clearScheduledStateChanges();
+    if (this.hbox.getAttribute("state") === "expanded") {
+      this.hbox.setAttribute("state", "collapsed");
+    }
+  }
+
+  _popupLearnMore() {
+    // TODO
+  }
+
+  _popupBlockRecommendations() {
+    // TODO
+  }
+
+  _popupInstallAddon() {
+    // TODO
+  }
+
+  /**
+   * Respond to a user click on the recommendation by showing a doorhanger/
+   * popup notification
+   */
+  _handleClick(event) {
+    const browser = this.window.gBrowser.selectedBrowser;
+    if (!RecommendationMap.has(browser)) {
+      // There's no recommendation for this browser, so the user shouldn't have
+      // been able to click
+      this.hide();
+      return;
+    }
+    const {message, name, icon, action} = RecommendationMap.get(browser);
+
+    // The recommendation should remain either collapsed or expanded while the
+    // doorhanger is showing
+    this._clearScheduledStateChanges();
+
+    // A hacky way of setting the popup anchor outside the usual url bar icon box
+    // See https://searchfox.org/mozilla-central/rev/847b64cc28b74b44c379f9bff4f415b97da1c6d7/toolkit/modules/PopupNotifications.jsm#42
+    browser.cfrpopupnotificationanchor = this.hbox;
+
+    let mainAction;
+    if (action.type === ACTION_TYPE_INSTALL_ADDON) {
```

Comment:
No need to check this if you use a generic dispatch handler (see commend below at line 150).